dxr/plugins/xpidl/visitor.py

```diff
@@ -0,0 +1,252 @@
+
+    foundinc = False
+    for inc in idl.includes():
+        if not foundinc:
+            foundinc = True
+            line += 1
+        line += number_lines(include % {'basename': idl_basename(inc.filename)})
+
+    if idl.needsJSTypes():
+        line += number_lines(jsvalue_include)
+
+    # Include some extra files if any attributes are infallible.
+    for iface in [p for p in idl.productions if p.kind == 'interface']:
+        for attr in [m for m in iface.members if isinstance(m, Attribute)]:
+            if attr.infallible:
+                line += number_lines(infallible_includes)
+                break
+
+    line += number_lines(header_end) + 1
+
+    for p in idl.productions:
+        production_map[p] = line
+        if p.kind == 'cdata':
+            line += number_lines(p.data)
+        elif p.kind == 'forward':
+            line += number_lines(forward_decl % {'name': p.name})
+        elif p.kind == 'interface':
+            # write_interface inserts a blank line at the start.
+            production_map[p] += 1
+            # Eh....
+            fd = StringIO()
+            write_interface(p, fd)
+            line += len(fd.readlines())
+        elif p.kind == 'typedef':
+            fd = StringIO()
+            printComments(fd, p.doccomments, '')
+            line += len(fd.readlines())
+            line += number_lines("typedef %s %s;\n\n" % (p.realtype.nativeType('in'),
+                                                         p.name))
+
+    return production_map
+
+
+class IdlVisitor(object):
+    """Traverse an IDL syntax tree and collect refs and needles."""
+
+    def __init__(self, parser, contents, rel_path, abs_path, include_folders, header_bucket,
+                 tree_config):
+        """
+        Parse an IDL file and visit the productions of its AST.
+
+        :arg parser: IdlParser to use to parse the file
+        :arg contents: the contents of the file, as a string
+        :arg rel_path: relative path to the file from tree root
+        :arg abs_path: absolute path to the file
+        :arg include_folders: list of folders to use to resolve include directives
+        :arg header_bucket: folder in which the generated header would be placed to make links
+        :arg tree_config: A :class:`TreeConfig` object for the current tree
+        """
+        self.tree = tree_config
+        # Hold on to the URL so we do not have to regenerate it everywhere.
+        self.header_filename = basename(rel_path.replace('.idl', '.h'))
+        header_path = relpath(join(header_bucket, self.header_filename),
+                              self.tree.source_folder)
+        self.generated_url = url_for('.browse',
+                                     tree=self.tree.name,
+                                     path=header_path)
+        ast = parser.parse(contents, basename(rel_path))
+        # Might raise IdlError
+        self.search_paths = [dirname(abs_path)] + include_folders
```

Comment:
It'd be nice to have comments here about the assumptions we're making. For example, here we seem to say "first we'll search the directory where an XPIDL file sits, and then we'll look at the explicitly provided include folders".